nmostesting/mocks/Auth.py

```diff
@@ -0,0 +1,230 @@
+                    format=serialization.PrivateFormat.TraditionalOpenSSL,
+                    encryption_algorithm=serialization.NoEncryption()
+                )
+                f.write(pem)
+
+    def make_metadata(self):
+        protocol = "http"
+        host = get_default_ip()
+        if ENABLE_HTTPS:
+            protocol = "https"
+            if DNS_SD_MODE == "multicast":
+                host = "nmos-mocks.local"
+            else:
+                host = "mocks.{}".format(DNS_DOMAIN)
+
+        metadata = {
+            "issuer": "{}://{}:{}".format(protocol, host, self.port),
+            "authorization_endpoint": "{}://{}:{}/auth".format(protocol, host, self.port),
+            "token_endpoint": "{}://{}:{}/token".format(protocol, host, self.port),
+            "jwks_uri": "{}://{}:{}/jwks".format(protocol, host, self.port),
+            "registration_endpoint": "{}://{}:{}/register".format(protocol, host, self.port),
+            "grant_types_supported": [
+                "authorization_code",
+                "implicit",
+                "refresh_token",
+                "password",
+                "client_credentials"
+            ],
+            "response_types_supported": [
+                "code",
+                "code token"
+            ],
+            "scopes_supported": [
+                "connection",
+                "node",
+                "query",
+                "registration",
+                "events",
+                "channelmapping"
+            ],
+            "code_challenge_methods_supported": [
+                "plain",
+                "S256"
+            ]
+        }
+        return metadata
+
+    def generate_token(self, scopes):
+        protocol = "http"
+        if ENABLE_HTTPS:
+            protocol = "https"
+            if DNS_SD_MODE == "multicast":
+                host = "nmos-mocks.local"
+            else:
+                host = "mocks.{}".format(DNS_DOMAIN)
+
+        pkey = open(self.key, "r").read()
+        token = generate_token(scopes, True, overrides={
+            "iss": "{}://{}:{}".format(protocol, host, self.port)}, private_key=pkey)
+        return token
+
+    def shutdown(self):
+        func = request.environ.get('werkzeug.server.shutdown')
+        if func is None:
+            raise RuntimeError('Not running with the Werkzeug Server')
+        func()
+
+
+AUTH = Auth()
+AUTH_API = Blueprint('auth', __name__)
+
+
+# special route to shutdown Auth Server
+@ AUTH_API.route('/shutdown', methods=['POST'])
+def shutdown():
+    AUTH.shutdown()
+    return 'Auth Server shutting down...'
+
+
+@ AUTH_API.route('/.well-known/oauth-authorization-server', methods=["GET"])
+def auth_metadata():
+    metadata = AUTH.make_metadata()
+    return Response(json.dumps(metadata), mimetype='application/json')
+
+
+@ AUTH_API.route('/jwks', methods=["GET"])
+def auth_jwks():
+    # get public key from RSA private key
+    pkey = open(AUTH.key, "r").read()
```

Comment:
Would it be possible to keep a copy of this in memory instead of read/write to disk?

Reply:
I kept cert and key in the file is because the loading SSLContext function is expecting cert and key files.

Reply:
I can see efforts to update SSLContext to support in-memory cert/key data stalled, see e.g. https://github.com/python/cpython/issues/60691, so I think we're left with using a read/writeable filesystem for now :-/